src/model/Enemy.hs

```diff
@@ -2,9 +2,99 @@ module Model.Enemy where
+                    | otherwise                  = b bullet 
+        spawn = BurstEnemy {(screenWidth, randomY) , burstEnemyCooldown}    
+
+instance Enemy ConeEnemy c where
+        hit c bullet | coneEnemyPos == bulletPos = Nothing
+                    | otherwise                 = c bullet 
+        spawn = ConeEnemy {(screenWidth, randomY) , coneEnemyCooldown}    
+
+instance Enemy BasicPlayerSeekingEnemy b where
+        hit b bullet | basicSeekingPos == bulletPos = Nothing 
+                    | otherwise                  = b bullet 
+        spawn = BasicPlayerSeekingEnemy {(screenWidth, randomY) , basicSeekingCooldown}      
+
+instance Enemy FastPlayerSeekingEnemy f where
+        hit f bullet | fastSeekingPos == bulletPos = Nothing 
+                    | otherwise                   = f bullet 
+        spawn = FastPlayerSeekingEnemy {(screenWidth, randomY) , fastSeekingCooldown}      
+
+       
+
+-- All enemies, except the player seeking types, are instances of this class.
+class MoveBasic a where 
+        move :: a -> a 
+    
+instance MoveBasic BasicEnemy b where
+        move b =  BasicEnemy {basicEnemyPos = (fst (basicEnemyPos b) - 1, snd (basicEnemyPos b))} -- Not completely sure how to change the tuple values of position.
+
+instance MoveBasic BurstEnemy b where
+        move b =  BurstEnemy {burstEnemyPos = (fst (burstEnemyPos b) - 1, snd (burstEnemyPos b))}
+
+instance MoveBasic ConeEnemy c where
+        move b =  ConeEnemy {coneEnemyPos = (fst (coneEnemyPos c) - 1, snd (coneEnemyPos c))} 
+
+
+    -- Only the two player seeking enemies are instances of this class.
+class MovePlayerSeeking where 
+        move :: a -> Player -> a        
+
+instance MovePlayerSeeking BasicPlayerSeekingEnemy b where
+        move b player =  BasicPlayerSeekingEnemy {basicSeekingPos = (fst (basicEnemyPos b) - 1, seeking_speed * abs (snd (basicSeekingPos b) - playerPos player))} -- The vertical movement is based on the vertical distance between player and enemy.
+                where seeking_speed = 0.25 -- How fast the enemy approaches the player vertically
+
+instance MovePlayerSeeking FastPlayerSeekingEnemy f where
+        move f player =  FastPlayerSeekingEnemy {fastSeekingPos = (fst (fastEnemyPos f) - 1, seeking_speed * abs (snd (fastSeekingPos f) - playerPos player))}
+                where seeking_speed = 0.45
+
+
+
+-- All enemies, except the fast player seeking type, are instances of this class.
+class CanShoot where 
+        shoot :: a -> [Bullets] 
+
+-- All bullet direction vectors
+left = (-1, 0)
+upward   = (-1, 1)
+downward = (-1, -1)
+
+instance CanShoot BasicEnemy b where
+        shoot b {basicEnemyPos} = [Bullet {basicEnemyPos, left}] -- One bullet
+                
+instance CanShoot BurstEnemy b where
+        shoot b {burstEnemyPos} = [Bullet {burstEnemyPos, left}, Bullet {burstEnemyPos - 1, left}, Bullet {burstEnemyPos - 2, left}] -- Three bullets in one horizontal line, two start a bit to the left of the enemy.
```

Comment:
Oh ja dit is wel een uitdaging... Ik denk dat we nu het risico gaan lopen dat de hit functie gaat activeren op die laatste twee bullets, aangezien die overlappen met de enemy zelf, en dat is niet wat we willen. Misschien is het makkelijk de voorste twee voor het schip te spawnen, en dat dan later weg te moffelen onder een animatie?

Reply:
Ja, goede oplossing, dus ze even soort van 'onzichtbaar' te maken met een animatie.